crates/swc_ecma_minifier/src/pass/mangle_names/mod.rs

```diff
@@ -1,40 +1,291 @@
+    fn lookup_char_pos(&self, _: BytePos) -> Loc {
+        unreachable!()
+    }
+
+    fn span_to_lines(&self, _: Span) -> Result<FileLines, SpanLinesError> {
+        unreachable!()
+    }
+
+    fn span_to_string(&self, _: Span) -> String {
+        String::new()
+    }
+
+    fn span_to_filename(&self, _: Span) -> FileName {
+        FileName::Anon
+    }
+
+    fn merge_spans(&self, _: Span, _: Span) -> Option<Span> {
+        None
+    }
+
+    fn call_span_if_macro(&self, sp: Span) -> Span {
+        sp
+    }
+
+    fn doctest_offset_line(&self, line: usize) -> usize {
+        line
+    }
+
+    fn span_to_snippet(&self, _: Span) -> Result<String, swc_common::SpanSnippetError> {
+        Ok(String::new())
+    }
+}
+
+impl SourceMapperExt for DummySourceMap {
+    fn get_code_map(&self) -> &dyn SourceMapper {
+        self
+    }
+}
+
+impl CharFreq {
+    pub fn scan(&mut self, s: &str, delta: i32) {
+        if delta == 0 {
+            return;
+        }
+
+        for c in s.chars() {
+            match c {
+                'a'..='z' => {
+                    self.0[c as usize - 'a' as usize] += delta;
+                }
+                'A'..='Z' => {
+                    self.0[c as usize - 'A' as usize + 26] += delta;
+                }
+                '0'..='9' => {
+                    self.0[c as usize - '0' as usize + 52] += delta;
+                }
+                '_' => {
+                    self.0[62] += delta;
+                }
+                '$' => {
+                    self.0[63] += delta;
+                }
+
+                _ => {}
+            }
+        }
+    }
+
+    pub fn compute(p: &Program, preserved: &FxHashSet<Id>) -> Self {
+        let mut buf = vec![];
+        let cm = Lrc::new(DummySourceMap);
+
+        {
+            let mut emitter = Emitter {
+                cfg: Default::default(),
+                cm,
+                comments: None,
+                wr: Box::new(JsWriter::new(Default::default(), "\n", &mut buf, None)),
+            };
+
+            emitter.emit_program(p).unwrap();
+        }
+
+        let code = String::from_utf8(buf).unwrap();
+
+        let mut freq = Self::default();
+
+        freq.scan(&code, 1);
+
+        // Subtract
+        p.visit_with(&mut CharFreqAnalyzer {
+            freq: &mut freq,
+            preserved,
+        });
+
+        freq
+    }
+
+    pub fn compile(self) -> Base54Chars {
+        static BASE54_DEFAULT_CHARS: &[u8; 64] =
+            b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ$_0123456789";
+
+        let mut arr = BASE54_DEFAULT_CHARS
+            .iter()
+            .copied()
+            .enumerate()
+            .map(|(idx, c)| (self.0[idx], c))
+            .collect::<Vec<_>>();
+
+        arr.sort_by_key(|&(freq, _)| Reverse(freq));
+
+        let mut head = vec![];
+        let mut tail = vec![];
+
+        for (_, c) in arr {
+            if !(b'0'..=b'9').contains(&c) {
+                head.push(c);
```

Comment:
Does this mean that `&head[..] != &tail[0..54]` ?
I am concerned that this will make the mapping of string and hashed number hard to calculate.

Reply:
Yes, I copied this logic from https://github.com/evanw/esbuild/blob/b2b5a602e4d0bcdb514dcc013d11be4dbf2cd3b3/internal/js_ast/js_ast.go#L1973-L1994

Reply:
I didn't think about it carefully, but maybe we can check/use logic of terser instead

Reply:
We perform string-based comparison instead.
https://github.com/kdy1/swc/blob/d13841a45fb79d31ea7bc24eb9d6c6a2cb55c103/crates/swc_ecma_minifier/src/pass/mangle_names/mod.rs#L248-L254